syntax = "proto3";

service NameNodeService {
    rpc RegisterDataNode (RegisterRequest) returns (RegisterResponse);
    rpc Heartbeat (HeartbeatRequest) returns (HeartbeatResponse);
    rpc AllocateBlocks (AllocateBlocksRequest) returns (AllocateBlocksResponse);
    rpc GetBlockLocations (BlockLocationRequest) returns (BlockLocationResponse);
    rpc GetFileBlocks (FileBlocksRequest) returns (FileBlocksResponse);
    rpc AddFile (AddFileRequest) returns (AddFileResponse);
    rpc ListFiles (ListFilesRequest) returns (ListFilesResponse);
    rpc Mkdir (MkdirRequest) returns (MkdirResponse);
    rpc Rmdir (RmdirRequest) returns (RmdirResponse);
    rpc RemoveFile (RemoveFileRequest) returns (RemoveFileResponse);
    rpc Move (MoveRequest) returns (MoveResponse); // Nuevo RPC para mv
    rpc Login (LoginRequest) returns (LoginResponse);
    rpc Logout (LogoutRequest) returns (LogoutResponse);
}

message RegisterRequest {
    string node_id = 1;
}
message RegisterResponse {
    bool success = 1;
}

message HeartbeatRequest {
    string node_id = 1;
}
message HeartbeatResponse {
    bool success = 1;
}

message AllocateBlocksRequest {
    string username = 1;
    int64 file_size = 2;
}
message AllocateBlocksResponse {
    repeated string block_ids = 1;
}

message BlockLocationRequest {
    string block_id = 1;
}
message BlockLocationResponse {
    repeated string node_ids = 1;
}

message FileBlocksRequest {
    string username = 1;
    string file_path = 2;
}
message FileBlocksResponse {
    repeated string block_ids = 1;
}

message AddFileRequest {
    string username = 1;
    string file_path = 2;
    repeated string block_ids = 3;
}
message AddFileResponse {
    bool success = 1;
    string file_path = 2; // Ruta del archivo en el DFS
}

message ListFilesRequest {
    string username = 1;
    string dir_path = 2;
}
message ListFilesResponse {
    repeated string items = 1;
}

message MkdirRequest {
    string username = 1;
    string dir_path = 2;
}
message MkdirResponse {
    bool success = 1;
}

message RmdirRequest {
    string username = 1;
    string dir_path = 2;
}
message RmdirResponse {
    bool success = 1;
}

message RemoveFileRequest {
    string username = 1;
    string file_path = 2;
}
message RemoveFileResponse {
    bool success = 1;
}

// --- Mensajes para Move (mv) ---
message MoveRequest {
    string username = 1;
    string source_path = 2;
    string destination_path = 3;
}

message MoveResponse {
    bool success = 1;
    string message = 2; // Mensaje opcional (ej. error o Ã©xito con detalles)
}

message LoginRequest {
    string username = 1;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
}

message LogoutRequest {
    string username = 1;
}

message LogoutResponse {
    bool success = 1;
    string message = 2;
}